packages/eslint-plugin/src/rules/prefer-optional-chain-utils/analyzeChain.ts

```diff
@@ -401,7 +409,42 @@ function getReportDescriptor(
     newCode = `!${newCode}`;
   }
 
-  const reportRange = getReportRange(chain, node.range, sourceCode);
+  const reportRange = getReportRange(chain, node, sourceCode);
+
+  if (node.type === AST_NODE_TYPES.IfStatement) {
+    const chainEndedWithSemicolon =
+      sourceCode.getTokenAfter(lastOperand.node)?.value === ';';
+    if (chainEndedWithSemicolon) {
+      newCode += ';';
+    }
+
+    const nodeBeforeTheComment = chainEndedWithSemicolon
+      ? lastOperand.node.parent
+      : lastOperand.node;
+    const commentsBefore = sourceCode.getCommentsBefore(chain[1].node);
+    const commentsAfter = sourceCode.getCommentsAfter(nodeBeforeTheComment);
```

Comment:
[Bug] It's hard to write comments handling code that's specific to types of comments & indentation. This very specific, which makes it brittle. For example:

* Other comments such as `#! env` exist
* Non-`' '` indentation exists, i.e. `\t`

I think you'll have a much better time using `sourceCode.getText`. As a reference, `prefer-function-type` slices text to capture everything in it:

https://github.com/typescript-eslint/typescript-eslint/blob/7ea2074ec74de1a4a9699ad96602b129c6f66762/packages/eslint-plugin/src/rules/prefer-function-type.ts#L109-L111

Does that kind of more general approach not work here?

Reply:
I kept a side branch where I tried something like that, but it got ugly real fast it almost doubled the code to handle comments and it still had issues with indentation.

It feels like the hardest part of lint fixers is dealing with comments because the tooling for text manipulation on AST nodes is so much better.

Reply:
Yeah 🥲 I really wish ASTs treated them the same...